File: src/ael/dev_oidc.py
"""Ephemeral OIDC-compatible issuer for local Compose acceptance only."""

from __future__ import annotations

import base64
import os
import time
from typing import Any

import jwt
from cryptography.hazmat.primitives.asymmetric import rsa
from fastapi import FastAPI

ISSUER = os.environ.get("AEL_DEV_OIDC_ISSUER", "http://dev-oidc:8080")
AUDIENCE = os.environ.get("AEL_DEV_OIDC_AUDIENCE", "ael-dev")
KEY_ID = "ael-development-ephemeral"
PRIVATE_KEY = rsa.generate_private_key(public_exponent=65537, key_size=2048)


def _b64uint(value: int) -> str:
    width = max(1, (value.bit_length() + 7) // 8)
    return base64.urlsafe_b64encode(value.to_bytes(width, "big")).decode().rstrip("=")


app = FastAPI(title="AEL development OIDC issuer")


@app.get("/.well-known/openid-configuration")
def discovery() -> dict[str, Any]:
    return {
        "issuer": ISSUER,
        "jwks_uri": f"{ISSUER}/.well-known/jwks.json",
        "token_endpoint": f"{ISSUER}/token",
        "id_token_signing_alg_values_supported": ["RS256"],
    }


@app.get("/.well-known/jwks.json")
def jwks() -> dict[str, Any]:
    numbers = PRIVATE_KEY.public_key().public_numbers()
    return {
        "keys": [
            {
                "kty": "RSA",
                "use": "sig",
                "alg": "RS256",
                "kid": KEY_ID,
                "n": _b64uint(numbers.n),
                "e": _b64uint(numbers.e),
            }
        ]
    }


@app.post("/token")
def token() -> dict[str, Any]:
    now = int(time.time())
    encoded = jwt.encode(
        {
            "iss": ISSUER,
            "aud": AUDIENCE,
            "sub": "ael-local-acceptance",
            "iat": now,
            "exp": now + 300,
        },
        PRIVATE_KEY,
        algorithm="RS256",
        headers={"kid": KEY_ID},
    )
    return {"access_token": encoded, "token_type": "Bearer", "expires_in": 300}


def main() -> None:
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8080)


if __name__ == "__main__":
    main()

File: internal/server/devoidc.go
package server

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"math/big"
	"net/http"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type DevOIDC struct {
	Issuer, Audience, Kid string
	key                   *rsa.PrivateKey
}

func NewDevOIDC(issuer, audience string) (*DevOIDC, error) {
	if issuer == "" || audience == "" {
		return nil, errors.New("development OIDC issuer and audience are required")
	}
	key, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, err
	}
	digest := sha256.Sum256(key.PublicKey.N.Bytes())
	return &DevOIDC{Issuer: strings.TrimRight(issuer, "/"), Audience: audience, Kid: hex.EncodeToString(digest[:8]), key: key}, nil
}
func (d *DevOIDC) Handler() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("GET /.well-known/openid-configuration", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, map[string]any{"issuer": d.Issuer, "jwks_uri": d.Issuer + "/.well-known/jwks.json", "token_endpoint": d.Issuer + "/token", "id_token_signing_alg_values_supported": []string{"RS256"}})
	})
	mux.HandleFunc("GET /.well-known/jwks.json", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, map[string]any{"keys": []any{map[string]string{"kty": "RSA", "use": "sig", "alg": "RS256", "kid": d.Kid, "n": base64.RawURLEncoding.EncodeToString(d.key.PublicKey.N.Bytes()), "e": encodeExponent(d.key.PublicKey.E)}}})
	})
	mux.HandleFunc("POST /token", d.token)
	mux.HandleFunc("GET /healthz", func(w http.ResponseWriter, r *http.Request) { writeJSON(w, 200, map[string]string{"status": "ready"}) })
	return mux
}
func (d *DevOIDC) token(w http.ResponseWriter, r *http.Request) {
	var request struct {
		Subject string `json:"subject"`
	}
	_ = json.NewDecoder(r.Body).Decode(&request)
	if request.Subject == "" {
		request.Subject = "aether-developer"
	}
	now := time.Now().UTC()
	claims := jwt.MapClaims{"iss": d.Issuer, "aud": d.Audience, "sub": request.Subject, "iat": now.Unix(), "exp": now.Add(15 * time.Minute).Unix()}
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	token.Header["kid"] = d.Kid
	signed, err := token.SignedString(d.key)
	if err != nil {
		writeError(w, 500, err)
		return
	}
	writeJSON(w, 200, map[string]any{"access_token": signed, "token_type": "Bearer", "expires_in": 900})
}
func encodeExponent(value int) string {
	bytes := big.NewInt(int64(value)).Bytes()
	return base64.RawURLEncoding.EncodeToString(bytes)
}
